### ISSUE TITLE
Multi-value fields not fully supported

### ISSUE THREAD

**Issue by joatca** [robertodoering/dart_mpd#1]
Thanks for this library! I hand-coded MPD stuff a while ago but this has the potential to make things a lot simpler. However I have need of proper support for multi-value tags. If I understand the current code:

https://github.com/robertodoering/dart_mpd/blob/master/lib/src/parser/parse_response.dart#L86

...multi-value fields are currently concatenated with a comma into a single value. I need to display each value separately so this fails when each tag could itself contain a comma; consider the Performer tags below:
```
OK MPD 0.23.5
currentsong
file: NAS/NAS/Baroque/Guitar/Guitar_Concertos_-_DECCA_464_0132/01.Concerto_in_D_major_for_4_guitars_-_1._Allegro-Vivaldi,_Antonio.flac
Last-Modified: 2022-01-23T18:26:15Z
Format: 44100:16:2
Performer: Los Romeros
Performer: Academy of St. Martin-in-the-Fields
Performer: Brown, Iona
Album: Guitar Concertos - DECCA 464 0132
Composer: Vivaldi, Antonio
Artist: Vivaldi, Antonio
Date: 1990
Title: Concerto in D major for 4 guitars - 1. Allegro
Genre: Baroque
Track: 1
Time: 184
duration: 184.106
Pos: 0
Id: 160
OK
```

Do you have any plans to add this and/or are you interesting in discussing the best way to achieve it?

**Comment by robertodoering** [comment]
Yes, the current approach just concatenates the values together, separated with a comma.
This worked for me but it's evidently not a good solution so I'd like to improve it.

My idea to fix this would be to parse the values into a `Map<String, dynamic>` where dynamic is either a `String` or a `List<String>`. This requires checking the type at runtime which is a bit annoying but it should be fine.

This would also mean the `MpdSong.tags` would be a `Map<String, dynamic>` and in your case you'd have a `List<String>` for the `Performer` tag.

Does that work for you or do you have a better idea?

**Comment by joatca** [comment]
To make it more typesafe I'd suggest parsing every tag into a `Map<String,List<String>>` (also available as `tags` I presume) and have that as the master format. Concatenation could be performed dynamically as required (and maybe cached, but that level of efficiency probably isn't needed).

We could have both singular and plural methods. Using my example again, `.performers()` could return the real data `["Los Romeros","Academy of St. Martin-in-the-Fields","Brown, Iona"]`, `.performer()` would return the list joined by the default separator `"Los Romeros,Academy of St. Martin-in-the-Fields,Brown, Iona"` and `performer(sep: ";")` would change the default separator and return `"Los Romeros;Academy of St. Martin-in-the-Fields;Brown, Iona"`

Obviously some attributes are single-value, like `format` and `file`. For simplicity they could be parsed in the same way, but the plural method would not be exposed, and the singular method could be defined more efficiently, like `tags["format"].isNotEmpty ? tags["format"].first : null`

My Dart experience is limited so apologies if any of this code isn't great.

**Comment by robertodoering** [comment]
That seems fine, but maybe a better approach that is still typesafe would be to use a union type with `freezed` for a single value and a list of values, which would look like this:
```dart
@freezed
class MpdValue with _$MpdValue {
  const factory MpdValue.single(String value) = MpdValueSingle;
  const factory MpdValue.multiple(List<String> values) = MpdValueMultiple;
}
```
We could also still expose a `toList()` method that always returns a `List<String>`.

Once dart3 is stable we can convert it to a sealed class so that we don't have to rely on code generation.

What do you think of that?

**Comment by joatca** [comment]
Yes, this seems better than my idea (although I don't really know how freezed works). Does this still allow the caller to choose the separator in the single case?

I like the name `multiple` when returning a list but I would suggest `concat` rather than `single` to make it clear what's happening if there are multiple values. `single` to me implies that it's more like `.multiple.first`.

I think it makes sense that `concat`/`single` should return `null` if there is no tag of that name; normally I hate nulls but they are much safer and easier to handle in Dart than in other languages so I actually prefer them. However what should `multiple` return if a tag doesn't exist? An empty list or `null`? It's more consistent to return `null` but I suspect that real life code is likely to loop over the returned values (and an empty list is therefore a no-op) so it would avoid constructs like `for (e in (foo.multiple ?? [])) {`. I guess I'm not experienced enough with Dart to know what the "Dart Way" is in this situation.

**Comment by robertodoering** [comment]
In that example `MpdValue` is a union type (or sealed class) where an instance is _either_ `MpdValueSingle` or `MpdValueMultiple`. So when you have a `MpdValue` instance, you have different ways to see whether it's a single value or multiple. Single is only the case when we got exactly one value for a key. Multiple when you have more than one value for a key.

E.g.
```dart
mpdValue.map(
  single: (value) {
    // value is a non-nullable string
  },
  multiple: (value) {
    // value is a non-nullable non-empty list of strings
  }
)
```

And with a `.toList()` method you can always get a List<String>. So when it's a single value it will just have a length of 1.

```dart
mpdValue.asList(); // always a List<String>
```

So I wouldn't concat values together anymore and we won't need to define a separator. If you wanted, you could get the same result that we have right now with `mpdValue.asList().join(',')`

I think this approach is the best to have the best of both worlds with a typesafe value that is dynamic based on whether we got only one or multiple values in the response.

**Comment by robertodoering** [comment]
Thanks for reporting the issue! I've pushed the changes to master, feel free to try them out.

There is also a different issue that also effect your `currentsong` response example, where we don't combine some values into one list and instead assume the duplicate tag indicates a new data set that is returned by mpd.
This includes arbitrary tags that we don't know about in the `currentsong` response.
I will be looking to fix this soon, but it requires a bit of rethinking on how the response is parsed.

This is how your example response is parsed right now:
```dart
{
  'file': MpdValue.single(value: NAS/NAS/Baroque/Guitar/Guitar_Concertos_-_DECCA_464_0132/01.Concerto_in_D_major_for_4_guitars_-_1._Allegro-Vivaldi,_Antonio.flac),
  'Last-Modified': MpdValue.single(value: 2022-01-23T18),
  'Format': MpdValue.single(value: 44100),
  'Performer': MpdValue.single(value: Los Romeros)
},
{
  'Performer': MpdValue.single(value: Academy of St. Martin-in-the-Fields)
},
{
  'Performer': MpdValue.single(value: Brown, Iona),
  'Album': MpdValue.single(value: Guitar Concertos - DECCA 464 0132),
  'Composer': MpdValue.single(value: Vivaldi, Antonio),
  'Artist': MpdValue.single(value: Vivaldi, Antonio),
  'Date': MpdValue.single(value: 1990),
  'Title': MpdValue.single(value: Concerto in D major for 4 guitars - 1. Allegro),
  'Genre': MpdValue.single(value: Baroque),
  'Track': MpdValue.single(value: 1),
  'Time': MpdValue.single(value: 184),
  'duration': MpdValue.single(value: 184.106),
  'Pos': MpdValue.single(value: 0),
  'Id': MpdValue.single(value: 160)
}

```

**Comment by joatca** [comment]
Sorry for the slow response. So if I understand correctly, in this case Freezed is emulating a union type. I actually don't think this is the correct way to handle the distinction between single and multiple values. I think it *does* make sense if we wanted to distinguish between, say, strings (title) and times (`Duration`) but not when distinguishing between a `String` and a `List<String>`. To me it makes more sense to treat a single value as a list of length 1 (and use `.join` as you suggest if you need to concatenate them to a single value). That is, everything is a list even though most of the time each one will have only one element. Even with `toList()`, representing single items differently makes the internal implementation unnecessarily complex.

The MPD protocol spec mentions that there may be more than one of each tag but it also calls out other metadata https://mpd.readthedocs.io/en/latest/protocol.html#other-metadata where there is only a single value, and for most of them you don't want a string, you want a `Duration` or a `DateTime` and a union type makes more sense here.

**Comment by robertodoering** [comment]
As a user of the library you only interact directly with `MpdValue` in the `MpdSong.tags` field - in all other models we can properly parse and type the value since they are not arbitrary.
For example, the other fields that are expected in the song response (duration, time, last-modified date etc.), are properly typed. See [MpdSong](https://github.com/robertodoering/dart_mpd/blob/master/lib/src/models/song.dart).

So the difference between a non-typesafe approach, a proper union class or always using a list is minor, but for me a union type makes much more sense and is nicer to work with.

**Comment by joatca** [comment]
Isn't `MpdSong` returning `String?` for the tags? Because there can be multiple values for at least artist and genre and MPD (and Vorbis tags from FLAC and Ogg) supports multiples for any tag AFAIKT.

**Comment by robertodoering** [comment]
Oh yeah you are right, vorbis does support multiple tags for 'Artist', 'Title' and 'Genre', where I wrongly assumed there can only be one.
I've now removed the `artist`, `title` and `genre` fields from MpdSong, so that they are now part of `MpdSong.tags` instead: https://github.com/robertodoering/dart_mpd/commit/df9091b52a528be872345a394107313243b56d0b

**Comment by robertodoering** [comment]
I've just published version 0.3.0 which now parses a song with multiple values per tag correctly.
Also, `MpdSong.tags` has been removed and all the tags are now parsed explicitly. E.g. you can use `MpdSong.performer`.
The tags are typed as `List<String>?` and the `MpdValue` union has been removed.
